docs: procedure to repair the DevWorkspace Operator (#2419)
codenvy/che-docs: modules/administration-guide/pages/repairing-the-devworkspace-operator.adoc

:_content-type: PROCEDURE
:navtitle: Repairing the {devworkspace} Operator
:description: Repairing the {devworkspace} Operator
:keywords: devworkspace, operator, repair
:page-aliases:

[id="repairing-the-devworkspace-operator_{context}"]
= Repairing the {devworkspace} Operator

.Prerequisites

* An active `{orch-cli}` session as a cluster administrator to the destination {orch-name} cluster. See {orch-cli-link}.

.Procedure

. Delete the `devworkspace-controller` namespace that contains the failing pod.

> . Update `DevWorkspace` and `DevWorkspaceTemplate` Custom Resource Definitions (CRD) by setting the conversion strategy to `None` and removing the next several lines of the removed strategy:

Reviewer: Sorry, overlooked this before.
```suggestion
. Update the `DevWorkspace` and `DevWorkspaceTemplate` Custom Resource Definitions (CRD) by setting the conversion strategy to `None` and removing the next several lines of the removed strategy:
```

Author: Thank you.

+
image::repairing-dwo/looking-up-custom-resource-definitions.png[]
+
[source,yaml]
----
spec:
  conversion:
    strategy: None # note .spec.conversion.webhook is removed here
----
+
image::repairing-dwo/setting-the-conversion-strategy-to-none.png[]
+
NOTE: The `DevWorkspaceOperatorConfig` and `DevWorkspaceRouting` CRDs  have the conversion strategy set to `None` by default.

. Remove the {devworkspace} Operator subscription:
+
----
$ oc delete sub devworkspace-operator
----

. Get the {devworkspace} Operator CSVs:
+
----
$ oc get csv | grep devworkspace
----
+
pass:[<!-- vale CheDocs.Attributes = NO -->]
.Output displayed in the __<devworkspace-operator.vX.Y.Z>__ format
====

> devworkspace-operator.v0.15.1                    DevWorkspace Operator   0.15.1            devworkspace-operator.v0.15.0                    Replacing
> devworkspace-operator.v0.15.2                    DevWorkspace Operator   0.15.2            devworkspace-operator.v0.15.1                    Pending

Author: ```suggestion
devworkspace-operator.v0.15.1    DevWorkspace Operator   0.15.1    devworkspace-operator.v0.15.0    Replacing
devworkspace-operator.v0.15.2     DevWorkspace Operator   0.15.2    devworkspace-operator.v0.15.1    Pending
```

:_content-type: PROCEDURE
:navtitle: Repairing the {devworkspace} Operator
:description: Repairing the {devworkspace} Operator
:keywords: devworkspace, operator, repair
:page-aliases:

[id="repairing-the-devworkspace-operator_{context}"]
= Repairing the {devworkspace} Operator

.Prerequisites

* An active `{orch-cli}` session as a cluster administrator to the destination {orch-name} cluster. See {orch-cli-link}.

.Procedure

. Delete the `devworkspace-controller` namespace that contains the failing pod.

> . Update `DevWorkspace` and `DevWorkspaceTemplate` CRDs by setting the conversion strategy to `None`:

Reviewer: ```suggestion
. Update `DevWorkspace` and `DevWorkspaceTemplate` Custom Resource Definitions (CRD) by setting the conversion strategy to `None`:
```

+

> image::repairing-dwo/looking-up-custom-resource-definitions.png[]

Reviewer: This is a file name, so it doesn't matter much, but still: don't use phrasal verbs in docs.

+
[source,yaml]
----
spec:
  conversion:
    strategy: None # note .spec.conversion.webhook is removed here
----
+

> image::repairing-dwo/setting-the-conversion-strategy-to-none.png[]

Reviewer: This image is redundant.

+
NOTE: The `DevWorkspaceOperatorConfig` and `DevWorkspaceRouting` CRDs  have the conversion strategy set to `None` by default.

. Remove the {devworkspace} Operator subscription:
+

> [source,terminal]

Reviewer: ```suggestion
```

----
$ oc delete sub devworkspace-operator
----

. Get the {devworkspace} Operator CSVs:
+

> [source,terminal]

Reviewer: ```suggestion
```

----
$ oc get csv | grep devworkspace
----
+
pass:[<!-- vale CheDocs.Attributes = NO -->]
.Output displayed in the __<devworkspace-operator.vX.Y.Z>__ format

> ====

Reviewer: Reduce the amount of white space to minimize the probability of the lines wrapping.

Author: Done.

devworkspace-operator.v0.15.1                    DevWorkspace Operator   0.15.1            devworkspace-operator.v0.15.0                    Replacing
devworkspace-operator.v0.15.2                    DevWorkspace Operator   0.15.2            devworkspace-operator.v0.15.1                    Pending
====
pass:[<!-- vale CheDocs.Attributes = YES -->]

. Remove each {devworkspace} Operator CSV:
+

> [source,terminal,subs="+quotes"]

Reviewer: ```suggestion
[subs="+quotes"]
```

----
$ oc delete csv __<devworkspace-operator.vX.Y.Z>__ <1>
----
<1> For example, `devworkspace-operator.v0.15.1` or `devworkspace-operator.v0.15.2`.

. Re-create the {devworkspace} Operator subscription:
+

> [source,terminal,subs="+attributes"]

Reviewer: ```suggestion
[subs="+attributes"]
```

----
$ cat <<EOF | oc apply -f -
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: devworkspace-operator
  namespace: openshift-operators
spec:
  channel: fast
  name: devworkspace-operator
  source: redhat-operators
  sourceNamespace: openshift-marketplace
  installPlanApproval: Manual
  startingCSV: devworkspace-operator.v{devworkspace-operator-version-patch}
EOF
----

> . Approve manual installation.

Reviewer: This seems unclear. How is it approved? Clicking some button (what button? where?), or something else?
```suggestion
. Approve the manual installation.
```

Author: I'm asking Ilya to clarify this.

> . Verify that manual installation has succeeded:

Reviewer: ```suggestion
. Verify that the manual installation has succeeded:
```

+

> image::repairing-dwo/verifying-success.png[]

Reviewer: Better to describe where to find the status than using an image (even from the image, it's not immediately clear where to look).

:_content-type: PROCEDURE
:navtitle: Repairing the {devworkspace} Operator
:description: Repairing the {devworkspace} Operator
:keywords: devworkspace, operator, repair
:page-aliases:

[id="repairing-the-devworkspace-operator_{context}"]
= Repairing the {devworkspace} Operator

> 

Reviewer 2: Missing a leading sentence explaining when a user should consider running this procedure: what are the symptoms of a failed {devworkspace} operator leading to the necessity of a repair?

Reviewer 2: And maybe: what are the most common causes leading to a broken operator?

Reviewer 3: Basically, the symptoms are 2 csv on the cluster in the 'Replacing' and 'Pending' states that are described below.

Author: I added a leading paragraph:
`
Under certain conditions, such as https://docs.openshift.com/container-platform/4.10/operators/understanding/olm/olm-understanding-olm.html[OLM] restart or cluster upgrade, the {prod2} Operator for {prod-short} might automatically install the {devworkspace} Operator even when it is already present on the cluster. In that case, you can repair the {devworkspace} Operator on OpenShift as follows:
`

.Prerequisites

* An active `{orch-cli}` session as a cluster administrator to the destination {orch-name} cluster. See {orch-cli-link}.

> 

Reviewer 2: Missing a prerequisite explaining when it may work: are the version of {prod-short} or {devworkspace} operator relevant?

Author: Like the rest of the docs content, it's for the product version and devworkspace operator version that are specified in antora.yml.

.Procedure

. Delete the `devworkspace-controller` namespace that contains the failing pod.

. Update `DevWorkspace` and `DevWorkspaceTemplate` Custom Resource Definitions (CRD) by setting the conversion strategy to `None` and removing the next several lines of the removed strategy:
+
[source,yaml]
----
spec:
  ...
  conversion:
    strategy: None # note .spec.conversion.webhook is removed here
status:
...
----
+

> TIP: You can find and edit the `DevWorkspace` and `DevWorkspaceTemplate` CRDs in the *Administrator* perspective of the OpenShift web console by searching for `DevWorkspace` in menu:Administration[CustomResourceDefinitions].

Reviewer 2: For consistency with the rest of the procedure and the docs, consider documenting how to achieve the result using the {orch-cli} command.

Reviewer 3: patching CRDs is a pretty delicate step, so it makes sense to do it from UI 
plus there are screenshots that exactly show which CRDs need to be patched and how exactly

Author: I checked the `oc` reference in the OpenShift docs, and it appears `oc` does not cover such functionality among its options.

+
NOTE: The `DevWorkspaceOperatorConfig` and `DevWorkspaceRouting` CRDs  have the conversion strategy set to `None` by default.

. Remove the {devworkspace} Operator subscription:
+
----

> $ oc delete sub devworkspace-operator

Reviewer 2: Is it OpenShift specific? If not, use `{orch-cli}` rather than `oc`.

Author: Confirmed by Ilya--OpenShift-specific. So I'm updating the heading to limit the scope to OpenShift.

----

. Get the {devworkspace} Operator CSVs:
+
----

> $ oc get csv | grep devworkspace

Reviewer 2: Is it OpenShift specific? If not, use `{orch-cli}` rather than `oc`.

Reviewer 2: Additionally, is it possible to have an output directly usable in the next step?

(And even better, concatenate both steps in one: find the CSVs and delete them).

Author: Confirmed by Ilya--OpenShift-specific. So I'm updating the heading to limit the scope to OpenShift.

Author: IMO, the output is complicated, and it's not worth the effort and time to try to write a script to merge these two steps into one. On top of that, the administrator better delete those manually.

----
+
pass:[<!-- vale CheDocs.Attributes = NO -->]

.Output displayed in the __<devworkspace-operator.vX.Y.Z>__ format
====
----
devworkspace-operator.v0.15.1    DevWorkspace Operator 0.15.1    devworkspace-operator.v0.15.0    Replacing

devworkspace-operator.v0.15.2    DevWorkspace Operator 0.15.2    devworkspace-operator.v0.15.1    Pending
----

> ====

Reviewer 2: Avoid using hard-coded version numbers, to avoid unnecessary maintenance overhead.

Use attributes for version numbers.

Even better, consider removing this example, as it is not necessary for understanding the procedure.

Reviewer 3: This is just an example so it should be ok

Author: I managed to incorporate the format into the sentence of the stop, and so I removed the example anyway.

pass:[<!-- vale CheDocs.Attributes = YES -->]

. Remove each {devworkspace} Operator CSV:
+
[subs="+quotes"]
----
$ oc delete csv __<devworkspace-operator.vX.Y.Z>__ <1>
----
<1> For example, `devworkspace-operator.v0.15.1` or `devworkspace-operator.v0.15.2`.

. Re-create the {devworkspace} Operator subscription:
+
[subs="+attributes"]
----
$ cat <<EOF | oc apply -f -
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: devworkspace-operator
  namespace: openshift-operators
spec:
  channel: fast
  name: devworkspace-operator
  source: redhat-operators
  sourceNamespace: openshift-marketplace
  installPlanApproval: Manual
  startingCSV: devworkspace-operator.v{devworkspace-operator-version-patch}
EOF
----

. In the *Administrator* perspective of the OpenShift web console, go to menu:Operators[Installed Operators] and select the following for the *{devworkspace} Operator*: menu:Upgrade available[Preview InstallPlan > Approve].

> . In the *Administrator* perspective of the OpenShift web console, go to menu:Operators[Installed Operators] and verify the *Succeeded* status of the *{devworkspace} Operator*.

Reviewer 2: For consistency, is it possible to use the `{orch-cli}` CLI for these last 2 steps?

Author: Confirmed by Ilya--OpenShift-specific. So I'm updating the heading to limit the scope to OpenShift.

Author: I checked the `oc` reference in the OpenShift docs, and it appears `oc` does not cover such functionality among its options.
